swiftsimio/writer.py

```diff
@@ -200,6 +200,120 @@ def write_particle_group(self, file_handle: h5py.File, compress: bool):
 
         return
 
+    def write_particle_group_metadata(
+        self, file_handle: h5py.File, dset_attributes: dict
+    ):
+        """
+        Writes the relevant metadata for a particle group
+
+        Parameters
+        ----------
+        file_handle: h5py.File
+            hdf5 output file being written
+        dset_attributes: dict
+            dictionary containg metadata to attach to group
+        """
+        for name, output_handle in getattr(
+            metadata.required_fields, self.particle_name
+        ).items():
+            obj = file_handle[f"/PartType{self.particle_type}/{output_handle}"]
+            for attr_name, attr_value in dset_attributes[output_handle].items():
+                obj.attrs.create(attr_name, attr_value)
+
+        return
+
+    def get_attributes(self, scale_factor: float) -> dict:
+        """
+        Returns a dictionary containg the attributes to attach to the dataset
+
+        Parameters
+        ----------
+        scale_factor: float
+            the cosmological scale factor of the dataset
+
+        Returns
+        -------
+        attributes_dict: dict
+            dictionary containg the attributes applying to the dataset
+        """
+        attributes_dict = {}
+        a_exp_dict = {
+            "coordinates": 1,
+            "internal_energies": -2,
+        }
+
+        for name, output_handle in getattr(
+            metadata.required_fields, self.particle_name
+        ).items():
+            field = getattr(self, name)
+
+            # Find the exponents for each of the dimensions
+            dim_exponents = get_dimensions(field.units.dimensions)
+
+            # Find the scale factor associated quantities
+            a_exp = a_exp_dict.get(name, 0)
+            a_factor = scale_factor ** a_exp
+
+            attributes_dict[output_handle] = {
+                "Conversion factor to CGS (not including cosmological corrections)": [
+                    field.unit_quantity.in_cgs()
+                ],
+                "Conversion factor to physical CGS (including cosmological corrections)": [
+                    field.unit_quantity.in_cgs() * a_factor
+                ],
+                # Assign the exponents in the proper order (see unyt.dimensions.base_dimensions)
+                "U_I exponent": [dim_exponents[5]],
+                "U_L exponent": [dim_exponents[1]],
+                "U_M exponent": [dim_exponents[0]],
+                "U_T exponent": [dim_exponents[3]],
+                "U_t exponent": [dim_exponents[2]],
+                "a-scale exponent": [a_exp],
+                "h-scale exponent": [0.0],
+            }
+
+        return attributes_dict
+
+
+def get_dimensions(dimension: unyt.dimensions) -> np.ndarray:
```

Comment:
I don't like this returning an array like this - can we make this a dictionary, or an object that it returns? Seeing as this is immutable maybe it should return a `namedtuple`.

Reply:
Changed this to a dictionary.